csrc/id_model/id_model.cpp

```diff
@@ -0,0 +1,283 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+
+#include <device_lower/analysis/trivial_broadcast.h>
+#include <device_lower/lower2device.h>
+#include <device_lower/utils.h>
+#include <disjoint_set.h>
+#include <ir/utils.h>
+#include <root_domain_map.h>
+#include <transform_iter.h>
+
+#include <tuple>
+#include <typeinfo>
+#include <utility>
+
+namespace nvfuser {
+
+namespace {
+
+// Map through loop swizzles, as input/output IterDomains are exact, only the
+// order they're traversed differs.
+void mapThroughLoopSwizzles(ValGraph& graph) {
+  std::vector<Swizzle2D*> all_swizzles;
+
+  for (const auto& expr_set :
+       std::as_const(graph).disjointExprSets().disjointSets()) {
+    auto swizzles_in_expr_set = ir_utils::filterByType<Swizzle2D>(
+        expr_set->vector().begin(), expr_set->vector().end());
+    all_swizzles.insert(
+        all_swizzles.end(),
+        swizzles_in_expr_set.begin(),
+        swizzles_in_expr_set.end());
+  }
+
+  for (auto swizzle : all_swizzles) {
+    if (swizzle->swizzleMode() == SwizzleMode::Loop) {
+      graph.mapVals(swizzle->inX(), swizzle->outX());
+      graph.mapVals(swizzle->inY(), swizzle->outY());
+    }
+  }
+}
+
+} // namespace
+
+IdModel::IdModel(
+    const std::vector<Expr*>& exprs,
+    const std::vector<TensorView*>& additional_tvs) {
+  build(exprs, additional_tvs);
+}
+
+IdModel::IdModel(const std::vector<Expr*>& exprs) : IdModel(exprs, {}) {}
+
+IdModel::IdModel(Fusion* fusion) {
+  std::vector<TensorView*> inputs_and_outputs;
+  {
+    auto inp_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), inp_tvs.begin(), inp_tvs.end());
+  }
+  {
+    auto out_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.end(), out_tvs.begin(), out_tvs.end());
+  }
+
+  build(fusion->exprs(), inputs_and_outputs);
```

Comment:
Here `inputs_and_outputs` are of the fusion. This is just to make sure all fusion inputs and outputs are included in the analysis, even when they are not used. For example, if a tensor is set as a fusion input but never used, we shouldn't need to bother with it but it might be more convenient to include them always as fusion inputs and outputs are not something we can just remove even if not used.